src/engine/Service.h

```diff
@@ -62,7 +79,15 @@ class Service : public Operation {
   // but in our tests (`ServiceTest`) we use a mock function that does not
   // require a running `HttpServer`.
   Service(QueryExecutionContext* qec, parsedQuery::Service parsedServiceClause,
-          GetResultFunction getResultFunction = sendHttpOrHttpsRequest);
+          NetworkFunctions networkFunctions = {
+              .getResultFunction_ = sendHttpOrHttpsRequest,
+              .getRuntimeInfoFunction_ = readHttpOrHttpsWebsocketStream});
+
+  ~Service() {
+    if (runtimeInfoThread_) {
+      runtimeInfoThread_->join();
+    }
```

Comment:
Use `ad_utility::JThread` , which joins itself on destruction if necessary and safe.